### ISSUE TITLE
Torch.autograd optimization is not working correctly when defined in Torch.autograd.function static method

### ISSUE THREAD

**Issue by daggertye** [pytorch/pytorch#32091]
## 🐛 Bug

When performing a task such as optimization within a Torch.autograd.function forward pass, no gradient functions are generated for intermittent non-leaf variables and hence optimization fails. 

## To Reproduce

Write out a custom torch.autograd.Function and attempt to perform some sort of optimization with a generated Parameter. Define any loss function with any basic operation and perform a loss.backward(). Sample code is shown below:

```
import torch
import torch.nn as nn


class ExampleFunction(torch.autograd.Function):
    @staticmethod
    def forward(ctx, x):
        y = nn.Parameter(torch.zeros_like(x))
        print(y)
        for i in range(100):
            if y.grad is not None:
                y.grad.detach_()
                y.grad.zero_()
            loss = torch.sum(x * y)
            print(loss)
            loss.backward()
            y.data -= 0.001 * y.grad

        return y

    @staticmethod
    def backward(ctx, grad_output):
        pass

ex = ExampleFunction.apply

print(ex(torch.ones(5)))
```

In the print out of the variables, we see that ```y``` is a parameter with ```requires_grad=True``` but ```loss``` does not have any ```grad_fn``` defined.

## Expected behavior

In the above example, ```loss``` should be a tensor with ```grad_fn``` defined. In particular, basic operations should propogate this grad_fn behavior.

## Environment

OS: Mac OSX 10.13.6
GCC version: Could not collect
CMake version: version 3.11.2

Python version: 3.6
Is CUDA available: No
CUDA runtime version: Could not collect
GPU models and configuration: Could not collect
Nvidia driver version: 1.1.0
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] msgpack-numpy==0.4.3.1
[pip] numpy==1.17.0
[pip] numpydoc==0.7.0
[pip] torch==1.3.1
[pip] torchvision==0.2.0
[conda] mkl                       2018.0.0             h5ef208c_6  
[conda] mkl-service               1.1.2            py36h7ea6df4_4  
[conda] torch                     1.3.1                     <pip>
[conda] torchvision               0.2.0                     <pip>

## Additional context

None

cc @ezyang @SsnL @albanD @zou3519 @gqchen

**Comment by albanD** [comment]
Hi,

This is expected as we do not perform automatic differentiation during the forward.
You can re-enable it for such use case with `with torch.enable_grad():` around your code that performs the local forward.

For future questions, you can use the forum: https://discuss.pytorch.org/
